src/Http/Routing/src/HeaderAttribute.cs

```diff
@@ -0,0 +1,81 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Microsoft.AspNetCore.Routing
+{
+    /// <summary>
+    /// Attribute for providing request header metdata that is used during routing.
+    /// </summary>
+    [DebuggerDisplay("{DebuggerToString(),nq}")]
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = false, Inherited = false)]
```

Comment:
Does it make sense to support multiple headers? Consider for example if you do versioning based on a header and want to have an endpoint that handles two versions

Reply:
You can do that with one header, just specify multiple acceptable values. What we cannot do today is require more than 1 header on the same route. We can extend support for that in the future.

Reply:
All of the other matcher policies allow multiple attributes.

* HTTP method policy allows you to have `[HttpGet]` and `[HttpPost]`.
* Host matcher policy allows you to have `[Host("*.contoso.com")]` and `[Host("contoso.com")]`.
* Consumes matcher policy allows multiple.

Instead of one attribute with a collection of values, why not multiple attributes, and have a single value? That would allow

```cs
[Header("x-version", "1")]
[Header("x-deprecated-version", "1")]
[Header("x-deprecated-version", "2")]
public ActionResult Whatever()
{
}
```

Reply:
Also, you can't assume there will only be one IHeaderMetadata. Specifying `AllowMultiple = false` on an attribute won't stop someone adding additional IHeaderMetadata to the endpoint when it is being built on startup.

Reply:
> What we cannot do today is require more than 1 header on the same route.

Base on what is commented out in YARP now, it needs multiple header key.

(https://github.com/microsoft/reverse-proxy/blob/master/src/ReverseProxy/Abstractions/RouteDiscovery/Contract/ProxyMatch.cs#L39)